native/src/components/TtsContainer.tsx

```diff
@@ -0,0 +1,194 @@
+import React, { createContext, ReactElement, useCallback, useEffect, useMemo, useState } from 'react'
+import { useTranslation } from 'react-i18next'
+import { AppState } from 'react-native'
+import Tts from 'react-native-tts'
+
+import buildConfig from '../constants/buildConfig'
+import { AppContext } from '../contexts/AppContextProvider'
+import { reportError } from '../utils/sentry'
+import TtsPlayer from './TtsPlayer'
+
+const unsupportedLanguagesForTts = ['fa', 'ka', 'kmr']
+
+export const isTtsActive = (content: string[] | null, languageCode: string): boolean =>
+  Array.isArray(content) &&
+  content.length > 0 &&
+  Boolean(buildConfig().featureFlags.tts) &&
+  !unsupportedLanguagesForTts.includes(languageCode)
+
+export type TtsContextType = {
+  visible: boolean
+  setVisible: (visible: boolean) => void
+  sentences: string[] | null
+  setSentences: (sentences: string[]) => void
+}
+
+export const ttsContext = createContext<TtsContextType>({
+  visible: false,
+  setVisible: () => undefined,
+  sentences: [],
+  setSentences: () => undefined,
+})
+
+type TtsContainerProps = {
+  children: ReactElement
+}
+
+const TtsContainer = ({ children }: TtsContainerProps): ReactElement => {
+  const { languageCode } = React.useContext(AppContext)
+  const { t } = useTranslation('layout')
+  const [isPlaying, setIsPlaying] = useState(false)
+  const [isExpanded, setIsExpanded] = useState(false)
+  const [sentenceIndex, setSentenceIndex] = useState(0)
+  const [visible, setVisible] = useState(false)
+  const [sentences, setSentences] = useState<string[]>([])
+
+  const unsupportedLanguage = unsupportedLanguagesForTts.includes(languageCode)
+  const maxTitle = 20
+  const title = sentences[0] || t('readAloud')
+  const longTitle = title.length > maxTitle ? title.substring(0, maxTitle).concat('...') : title
+
+  const callOnEnd = () => {
+    Tts.stop()
+    setIsPlaying(false)
+    setIsExpanded(false)
+    setSentenceIndex(0)
+  }
+
+  const initializeTts = useCallback((): void => {
+    Tts.getInitStatus().catch(async error => {
+      reportError(`Tts-Error: ${error.code}`)
+      if (error.code === 'no_engine') {
+        await Tts.requestInstallEngine().catch((e: string) => reportError(`Failed to install tts engine: : ${e}`))
+      }
+    })
+  }, [])
+
+  const startReading = (index = sentenceIndex) => {
+    if (!unsupportedLanguage && sentences.length > 0) {
+      // Persian not supported
+      Tts.setDefaultLanguage(languageCode)
+      Tts.speak(String(sentences[index]), {
+        androidParams: {
+          KEY_PARAM_PAN: 0,
+          KEY_PARAM_VOLUME: 0.6,
+          KEY_PARAM_STREAM: 'STREAM_MUSIC',
+        },
+        iosVoiceId: '',
+        rate: 1,
+      })
+      if (sentences.length === 0) {
+        setSentenceIndex(0)
+        setIsExpanded(false)
+      }
+    } else {
+      setIsPlaying(false)
+    }
```

Comment:
:x: I think responsibilities are currently a little unclear. This method should never have to set isPlaying to false again. Instead, it should only this method should set isPlaying to true at all (and only if there is actually something to play).
In my opinion, there should be the following methods/callback handlers:
- play
- pause
- stop (and/or close)
- playNext
- playPrevious

Each one of those should handle setting the correct state and edge cases (e.g. in the case of goForward, play should not be called if there is no next sentence).
So `play` would look something like:
``` ts
const play = (index = sentenceIndex) => {
  Tts.stop() // or custom stop() if necessary
  const sentence = sentences[index]
  if (sentence) {
    setIsPlaying(true)
    Tts.speak(...)
  }
}
```
while `playNext` would look something like this:
``` tsx
const newIndex = index + 1
if (nextIndex < sentences.length) {
  setSentenceIndex(nextIndex)
  play(nextIndex)
}
```
and similarly for the other functions.

Reply:
Sorry for overcomplicating things 😓

Reply:
No worries! I think you are doing a great job :)